### ISSUE TITLE
The provider creates mixed up SSH keys on new instances

### ISSUE THREAD

**Issue by multani** [exoscale/terraform-provider-exoscale#119]
I'm trying to migrate a project from the Cloudstack provider to the Exoscale provider.

Something that I did was to create a "exoscale" module in my project, which creates several specific instances in a specific Exoscale zone that I can pass as a variable. With some help, I've managed so far to create infrastructure on several zones at once, just by instantiating several times this "exoscale" module, with different "zone" values.

I quickly faced the problem of "how to deal with cross-zones objects?" (things which I've called "global objects", such as SSH keys, security groups, etc.).
With the Cloudstack provider, I was prefixing the name of these objects with a different prefix for every instances of my exoscale module, something like:
```
module "exoscale_frankfurt" {
  source = "modules/exoscale-cs"
  exoscale_zone       = "de-fra-1"
  globals_prefix      = "exo/fra"
  ssh_public_key_path = "~/.ssh/id_rsa.pub"
}

module "exoscale_vienna" {
  source = "modules/exoscale-cs"
  exoscale_zone       = "at-vie-1"
  globals_prefix      = "exo/vie"
  ssh_public_key_path = "~/.ssh/id_rsa.pub"
}
```

Note that I'm using the same SSH key in both instance: they are both created on Exoscale, using different names (prefix with the content of ``globals_prefix``), although the content and the SSH fingerprint are the same. This used to work pretty well with the Cloudstack provider.

However, after migrating this module over the Exoscale provider, if I reuse the example above, instances created in the Vienna zone end up having the Frankfurt SSH key. On subsequent apply, Terraform detects the configuration of the VM is not right and tries to recreate the instance.

You can see in [this Git repository](https://github.com/multani/exoscale-terraform-demo) how to reproduce the problem. There's the former cloudstack provider and the newer exoscale provider, with the same configuration. You will need to run ``terraform apply`` twice and the second time, you should see the SSH keys mixed.

I tried to debug the calls made on the Exoscale API by egoscale, it looks like the different Terraform tools ultimately call the endpoint with the right settings.

**Comment by greut** [comment]
Hi @multani,

there is an issue into Cloudstack itself where a key is indexed using its fingerprint rather than name. You may create many same ssh keys with different name. But when you retrieve the information from an instance name of the SSH keys might be the wrong one.

One solution is to ignore any changes to the key_pair.

```hcl
  lifecycle {
    ignore_changes = ["key_pair"]
}
```

An alternative approach would be to send the SSH keys via the _cloud-init_ config.

http://cloudinit.readthedocs.io/en/latest/topics/modules.html?highlight=ssh_authorized_keys#ssh

You lose, in the portal the information about which SSH key is installed, but you're now able to send more than one key.

A possible fix on this side would be to never override the SSH key with the value coming from CloudStack or perform an extra SSH call to figure out if it, in fact, changed.

Keep us updated on what you consider be best for your use case.

Best,

**Comment by multani** [comment]
Hi @greut,

Thanks for the propositions, another one would be to create the SSH key out of the modules (and only once then) and passes its ID down into each instance of the modules, so the same public key would not be duplicated.

What really puzzled me though is that I've created many instances in the past using the CloudStack provider and it never asked me to recreate or to update the SSH key associated with the instances.

So I just had a quick comparison between the CloudStack provider and Exoscale provider code and actually, the CloudStack provider never re-reads the key name set on a machine so it can't detect the one I've specified is different from the one I've set in my Terraform configuration. This is actually configured on Exoscale portal, where I just saw instances created via CloudStack have the wrong SSH key associated.

So it looks this provider has direct built-in support for the ``ignore_changes`` solution you proposed above :)

Thanks for the information @greut, now I understand better what's happening (on CloudStack) and why there's a difference between both providers!

**Comment by greut** [comment]
> So I just had a quick comparison between the CloudStack provider and Exoscale provider code and actually, the CloudStack provider never re-reads the key name set on a machine so it can't detect the one I've specified is different from the one I've set in my Terraform configuration.

It makes sense. Calls like [resetSSHKeyForVirtualMachine](http://cloudstack.apache.org/api/apidocs-4.4/user/resetSSHKeyForVirtualMachine.html) won't do much once the machine has been booted. Maybe used in combination with [restoreVirtualMachine](http://cloudstack.apache.org/api/apidocs-4.4/user/restoreVirtualMachine.html). Which is the only way to ensure you won't loose your IP address, but it's a bit of extra work.

**Comment by multani** [comment]
Actually, the CloudStack provider correctly detects the key changed, only if the configuration in Terraform is different from the state previously stored by Terraform (it does the diff at this level, but it doesn't read out the value from the CloudStack API directly), AFAIK.

In that case, it seems it correctly does a trigger a restart of the machine, although I haven't paid attention to all the details (like keeping the IP address as you mentioned.)
